mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala

```diff
@@ -408,8 +409,12 @@ class IsotonicRegression private (private var isotonic: Boolean) extends Seriali
    */
   private def parallelPoolAdjacentViolators(
       input: RDD[(Double, Double, Double)]): Array[(Double, Double, Double)] = {
-    val parallelStepResult = input
-      .sortBy(x => (x._2, x._1))
+    val keyedInput = input
```

Comment:
I think there may be shorter ways to write this with `groupBy`, but, this and other approaches like that have the big drawback of reading lots of data into memory. Here you have to sort the whole partition in memory (!).

How about `repartitionAndSortWithinPartitions`? oddly specific method, but, likely just what you need here, to both partition according to some criteria but then end up with sorted partitions. It's more scalable.